fdbserver/VersionedBTree.actor.cpp

```diff
@@ -2227,39 +2265,60 @@ class DWALPager final : public IPager2 {
 		if (exists && fileSize >= (self->smallestPhysicalBlock * 2)) {
 			debug_printf("DWALPager(%s) recovering using existing file\n", self->filename.c_str());
 
-			state bool recoveredHeader = false;
-
-			// Read physical page 0 directly
-			wait(store(self->headerPage, self->readHeaderPage(self, 0)));
-
-			// If the checksum fails for the header page, try to recover committed header backup from page 1
-			if (!self->headerPage->verifyChecksum(0)) {
-				TraceEvent(SevWarn, "RedwoodRecoveringHeader").detail("Filename", self->filename);
+			state bool recoveredBackupHeader = false;
 
-				wait(store(self->headerPage, self->readHeaderPage(self, 1)));
+			// Try to read primary header
+			try {
+				wait(store(self->headerPage, self->readHeaderPage(primaryHeaderPageID)));
+			} catch (Error& e) {
+				debug_printf("DWALPager(%s) Primary header read failed with %s\n", self->filename.c_str(), e.what());
+
+				// Errors that can be caused by a corrupted and unsync'd write to the header page can be ignored and the
+				// committed, sync'd backup header will be tried. Notably, page_header_wrong_page_id is not included in
+				// this list because it means the header checksum passed but this page data is not meant to be at this
+				// location, which is a very serious error so recovery should not continue.
+				bool tryBackupHeader =
+				    (e.code() == error_code_page_header_version_not_supported ||
+				     e.code() == error_code_page_encoding_not_supported ||
+				     e.code() == error_code_page_decoding_failed || e.code() == error_code_page_header_checksum_failed);
+
+				TraceEvent(SevWarn, "RedwoodRecoveryErrorPrimaryHeaderFailed")
+				    .error(e, true)
+				    .detail("Filename", self->filename)
+				    .detail("PageID", primaryHeaderPageID)
+				    .detail("TryingBackupHeader", tryBackupHeader);
 
-				if (!self->headerPage->verifyChecksum(1)) {
-					if (g_network->isSimulated()) {
-						// TODO: Detect if process is being restarted and only throw injected if so?
-						throw io_error().asInjectedFault();
-					}
+				// Don't throw if trying backup header below
+				if (!tryBackupHeader) {
```

Comment:
are there cases where we don't want to explicitly try reading backup header?

Reply:
No, but it's an interesting topic.

The disk contract used by simulation is very weak.  It is such that any writes, whether acknowledged or not, that are not fsync'd before a simulated reboot can have corrupt data written to their destination on disk which will then be seen upon read after the restart.  This is the contract Redwood is written to.  However, it's actually weaker than many modern block devices.

In theory, if by contract your disk device supported atomic update of the exact location and size of the space that holds the primary header page, then if a primary header comes back corrupt it means either
- Redwood issued a write of a bad header page at some point
- The disk device broke its contract

In this scenario it is tempting to say something bad enough has happened that recovery should stop.  However, it isn't an error to continue even if it means losing the latest commit, because upon recovery the StorageServer reads the durable version from the recovered storage engine and starts reading the log system from that point.  If the log system still has that version, all is well.  If it doesn't, the StorageServer is removed.  Either way, no incorrect behavior occurs by recovering from the backup header.